README.md
####工具说明：

liveshell工具是采用Wilddog云以及WildDog C/嵌入式SDK制作的，远程调用shell脚本的工具。它监听某个url下数据的变化，当数据发生变化后，触发预先安排好的可执行程序或者shell脚本，同时将变化后的数据作为参数传入。

WildDog C/嵌入式的SDK 使用的是CoAP UDP + DTLS + CBOR技术。

Linux发行版均可使用该工具，已在以下发行版中测试过：

	Linux版本                                                       内核版本
	centos-release-7-1.1503.el7.centos.2.8.x86_64                   3.10.0-229.el7.x86_64
	centos-release-6-5.el6.centos.11.1.x86_64                       2.6.32-431.el6.x86_64
	centos-release-5.5.el5.centos                                   2.6.18-194.el5
	Red Hat Enterprise Linux Server release 7.1  (Maipo)            3.10.0-229.el7.x86_64
	Red Hat Enterprise Linux Server release 6.6 (Santiago)          2.6.32-504.el6.x86_64
	Red Hat Enterprise Linux Server release 5.11 (Tikanga)          2.6.18-398.el5
	Ubuntu14.04.01                                                  3.16.0-30-generic
	Ubuntu12.04                                                     3.13.0-32-generic

除了上述各大主流的Linux发行版可以使用，在Openwrt系统中同样可以使用liveshell工具，具体的安装步骤请参见`docs`下的`LiveShell_used_in_Openwrt.md`文档。

####linux环境下安装说明：
#####1. 下载

从git下载到文件夹

	git clone https://github.com/WildDogTeam/liveshell.git
	
更新C/嵌入式SDK

	cd liveshell	
	git submodule update --init --recursive

#####2. 安装
注意权限，install需要root权限

	./configure
	make
	make install

#####3.使用

	liveshell  [option] <your wilddog url> <your callback command>

#####4. option参数说明

	-o 获取到原始数据也会触发一次command回调

	-s 当数据为字符串时，不解析该字符串，完整的将该字符串（以双引号包裹）传递给command回调
	
	-v 触发回调时将命令完整打出
	
	-h 打印帮助信息

	--version 打印版本号

	--authvalue=<your auth data> 传入auth数据（如超级密钥、token等）
	
#####5. 例子

######显示根目录的文件

1. 在Wilddog云端建立一颗数据树，结构为

		{
		  "path": "/"
		}	

2. 终端运行liveshell，将`<your Appid>`替换成你自己的appid

		liveshell -ov coaps://<your Appid>.wilddogio.com/path ls

3. 终端的显示结果等同于在终端运行"ls /"，修改云端数据，把"/"改为"-l /"，在终端立刻显示"ls -l /"的结果；类似的，你也可以将<your callback command>改为shell脚本，执行更加复杂的操作。

4. 在传入auth数据的情况下，我们可以将云端安全规则中的读写权限配置为false，这样只有在auth正确的条件下，才可以读写。

		liveshell -ov --authvalue=XXXXXXXEFAHYAUDAF1SACfEXXXXXXX coaps://<your Appid>.wilddogio.com/path ls


NEWS
Version 1.4 - Oct 22, 2015

* Add auth.

Version 1.3 - Oct 20, 2015

* Update submodule and make liveshell could be used in openwrt.

Version 1.2 - Oct 13, 2015

* Change submodule from wilddog_client_coap to wilddog-client-c.

Version 1.1 - Oct 10, 2015

* Cmd and string value support more than one arg now.

Version 1.0 - Oct 9, 2015, by Baikal Hu

* Initial commit, using Wilddog cloud and Wilddog C/Embedded SDK, developed 
an useful tool to remote call shell commands.


config.h
/* config.h.  Generated from config.h.in by configure.  */
/* config.h.in.  Generated from configure.in by autoheader.  */

/* Define to 1 if you have the `alarm' function. */
#define HAVE_ALARM 1

/* Define to 1 if you have the <arpa/inet.h> header file. */
#define HAVE_ARPA_INET_H 1

/* Define to 1 if you have the <fcntl.h> header file. */
#define HAVE_FCNTL_H 1

/* Define to 1 if you have the `fork' function. */
#define HAVE_FORK 1

/* Define to 1 if you have the `gethostbyname' function. */
#define HAVE_GETHOSTBYNAME 1

/* Define to 1 if you have the `gettimeofday' function. */
#define HAVE_GETTIMEOFDAY 1

/* Define to 1 if you have the <inttypes.h> header file. */
#define HAVE_INTTYPES_H 1

/* Define to 1 if your system has a GNU libc compatible `malloc' function, and
   to 0 otherwise. */
#define HAVE_MALLOC 1

/* Define to 1 if you have the <malloc.h> header file. */
#define HAVE_MALLOC_H 1

/* Define to 1 if you have the `memmove' function. */
#define HAVE_MEMMOVE 1

/* Define to 1 if you have the <memory.h> header file. */
#define HAVE_MEMORY_H 1

/* Define to 1 if you have the `memset' function. */
#define HAVE_MEMSET 1

/* Define to 1 if you have the <netdb.h> header file. */
#define HAVE_NETDB_H 1

/* Define to 1 if you have the <netinet/in.h> header file. */
#define HAVE_NETINET_IN_H 1

/* Define to 1 if the system has the type `ptrdiff_t'. */
#define HAVE_PTRDIFF_T 1

/* Define to 1 if your system has a GNU libc compatible `realloc' function,
   and to 0 otherwise. */
#define HAVE_REALLOC 1

/* Define to 1 if you have the `select' function. */
#define HAVE_SELECT 1

/* Define to 1 if you have the `socket' function. */
#define HAVE_SOCKET 1

/* Define to 1 if you have the <stddef.h> header file. */
#define HAVE_STDDEF_H 1

/* Define to 1 if you have the <stdint.h> header file. */
#define HAVE_STDINT_H 1

/* Define to 1 if you have the <stdlib.h> header file. */
#define HAVE_STDLIB_H 1

/* Define to 1 if you have the `strcasecmp' function. */
#define HAVE_STRCASECMP 1

/* Define to 1 if you have the `strchr' function. */
#define HAVE_STRCHR 1

/* Define to 1 if you have the <strings.h> header file. */
#define HAVE_STRINGS_H 1

/* Define to 1 if you have the <string.h> header file. */
#define HAVE_STRING_H 1

/* Define to 1 if you have the `strstr' function. */
#define HAVE_STRSTR 1

/* Define to 1 if you have the <sys/socket.h> header file. */
#define HAVE_SYS_SOCKET_H 1

/* Define to 1 if you have the <sys/stat.h> header file. */
#define HAVE_SYS_STAT_H 1

/* Define to 1 if you have the <sys/time.h> header file. */
#define HAVE_SYS_TIME_H 1

/* Define to 1 if you have the <sys/types.h> header file. */
#define HAVE_SYS_TYPES_H 1

/* Define to 1 if you have the <unistd.h> header file. */
#define HAVE_UNISTD_H 1

/* Define to 1 if you have the `vfork' function. */
#define HAVE_VFORK 1

/* Define to 1 if you have the <vfork.h> header file. */
/* #undef HAVE_VFORK_H */

/* Define to 1 if `fork' works. */
#define HAVE_WORKING_FORK 1

/* Define to 1 if `vfork' works. */
#define HAVE_WORKING_VFORK 1

/* Define to 1 if the system has the type `_Bool'. */
#define HAVE__BOOL 1

/* Name of package */
#define PACKAGE "LiveShell"

/* Define to the address where bug reports for this package should be sent. */
#define PACKAGE_BUGREPORT "BUG-REPORT-ADDRESS"

/* Define to the full name of this package. */
#define PACKAGE_NAME "FULL-PACKAGE-NAME"

/* Define to the full name and version of this package. */
#define PACKAGE_STRING "FULL-PACKAGE-NAME VERSION"

/* Define to the one symbol short name of this package. */
#define PACKAGE_TARNAME "full-package-name"

/* Define to the home page for this package. */
#define PACKAGE_URL ""

/* Define to the version of this package. */
#define PACKAGE_VERSION "VERSION"

/* Define to 1 if you have the ANSI C header files. */
#define STDC_HEADERS 1

/* Version number of package */
#define VERSION "1.0"

/* des */
/* #undef WILDDOG_LITTLE_ENDIAN */

/* des */
/* #undef WILDDOG_MACHINE_BITS */

/* Description */
#define WILDDOG_PORT 5684

/* des */
/* #undef WILDDOG_PROTO_MAXSIZE */

/* des */
/* #undef WILDDOG_RECEIVE_TIMEOUT */

/* des */
/* #undef WILDDOG_REQ_QUEUE_NUM */

/* des */
/* #undef WILDDOG_RETRANSMITE_TIME */

/* Define for Solaris 2.5.1 so the uint32_t typedef from <sys/synch.h>,
   <pthread.h>, or <semaphore.h> is not used. If the typedef were allowed, the
   #define below would cause a syntax error. */
/* #undef _UINT32_T */

/* Define for Solaris 2.5.1 so the uint64_t typedef from <sys/synch.h>,
   <pthread.h>, or <semaphore.h> is not used. If the typedef were allowed, the
   #define below would cause a syntax error. */
/* #undef _UINT64_T */

/* Define for Solaris 2.5.1 so the uint8_t typedef from <sys/synch.h>,
   <pthread.h>, or <semaphore.h> is not used. If the typedef were allowed, the
   #define below would cause a syntax error. */
/* #undef _UINT8_T */

/* Define to `__inline__' or `__inline' if that's what the C compiler
   calls it, or to nothing if 'inline' is not supported under any name.  */
#ifndef __cplusplus
/* #undef inline */
#endif

/* Define to the type of a signed integer type of width exactly 16 bits if
   such a type exists and the standard includes do not define it. */
/* #undef int16_t */

/* Define to the type of a signed integer type of width exactly 32 bits if
   such a type exists and the standard includes do not define it. */
/* #undef int32_t */

/* Define to the type of a signed integer type of width exactly 64 bits if
   such a type exists and the standard includes do not define it. */
/* #undef int64_t */

/* Define to rpl_malloc if the replacement function should be used. */
/* #undef malloc */

/* Define to `int' if <sys/types.h> does not define. */
/* #undef pid_t */

/* Define to rpl_realloc if the replacement function should be used. */
/* #undef realloc */

/* Define to `unsigned int' if <sys/types.h> does not define. */
/* #undef size_t */

/* Define to `int' if <sys/types.h> does not define. */
/* #undef ssize_t */

/* Define to the type of an unsigned integer type of width exactly 16 bits if
   such a type exists and the standard includes do not define it. */
/* #undef uint16_t */

/* Define to the type of an unsigned integer type of width exactly 32 bits if
   such a type exists and the standard includes do not define it. */
/* #undef uint32_t */

/* Define to the type of an unsigned integer type of width exactly 64 bits if
   such a type exists and the standard includes do not define it. */
/* #undef uint64_t */

/* Define to the type of an unsigned integer type of width exactly 8 bits if
   such a type exists and the standard includes do not define it. */
/* #undef uint8_t */

/* Define as `fork' if `vfork' does not work. */
/* #undef vfork */


src/liveshell.c
/*
 * Copyright (C) 2014-2016 Wilddog Technologies. All Rights Reserved. 
 *
 * FileName: liveshell.c
 *
 * Description: liveshell file.
 *
 * Usage: liveshell    [option] <your wilddog url> 
 *                      <your callback command> 
 *      
 *
 * History:
 * Version      Author          Date                    Description
 *
 * 1.0.0        Baikal.Hu       2015-09-28              Create file.
 *
 */

#include <unistd.h>
#include <stdio.h>
#include <stdlib.h>
#include <getopt.h>
#include <string.h> 
#include <sys/types.h>
#include "wilddog.h"


#define  LIVESHELL_VERSION	"1.4" 

STATIC void auth_callback
    (
        void* arg,
        Wilddog_Return_T err
    )
{
    *(BOOL*)arg = TRUE;
    if(err < WILDDOG_HTTP_OK || err >= WILDDOG_HTTP_NOT_MODIFIED)
    {
        wilddog_debug("auth failed!;error =%d \n",err);
        return;
    }
}


Wilddog_Str_T * getHostFromUrl(Wilddog_Str_T *url)
{
	int length = 0;
	int m = 0, n = 0;
	int i = 0;
	Wilddog_Str_T *host = NULL;

	length = strlen(url);
	for(i = 0; i < length; i++)
	{
		if(*(url+i) == '/' && *(url+i+1) == '/')
			break;
	}

	m = i+2;

	for(i = m; i < length; i++)
	{
		if(*(url+i) == '/')
			break;
	}
	n = i-1;

	host = wmalloc(n-m+2);
	if(NULL == host)
	{
		
	}

	strncpy(host, url+m, n-m+1);
	return host;
}


typedef struct _watch_ctx
{
    BOOL *isFinished;
    char *cmd;
    char *data;
    char *verbose;
    char *quote;
}WATCH_CTX;

STATIC void observer_callback
    (
    const Wilddog_Node_T* p_snapshot, 
    void* arg,
    Wilddog_Return_T err
    )
{
    int ret = 0;
    char cmd[1024];
    static int times=0;
    pid_t pid;
    char * pstr = NULL;
    char * pleaf = NULL;
    Wilddog_Node_T *p_clone = NULL;
    Wilddog_Str_T *p_key = NULL;
    char *d = " ";
    char *ps = NULL;
    char *args[0];
    int i = 0,j = 1;

    memset(cmd,0,1024);
    *((*(WATCH_CTX*)arg).isFinished)= TRUE;
    if(err < WILDDOG_HTTP_OK || err >= WILDDOG_HTTP_NOT_MODIFIED)
    {
        wilddog_debug("addObserver failed!");
        return;
    }

    if(strncmp("yes", (*(WATCH_CTX*)arg).data, strlen("yes")) == 0 && times == 0)
    {
        times++;
        return;
    }

    pid = fork();
    if(pid == -1)
        perror("fork");

    if(!pid)
    {
        p_clone = wilddog_node_clone(p_snapshot);
        p_key = p_clone->p_wn_key;
        p_clone->p_wn_key = NULL;
        pstr = (char *)wilddog_debug_n2jsonString(p_clone);
        fflush(stdout);
        /*Handle cmd's arg, add to arg table. Such as "ls -l", "-l" is an arg*/
        
        if((strncmp("no", (*(WATCH_CTX*)arg).quote, strlen("no")) != 0) && p_clone->p_wn_child == NULL && p_clone->d_wn_type == WILDDOG_NODE_TYPE_UTF8STRING)
        {
            /*If node is leaf and value is string, seperate to args by space*/
            pleaf = wmalloc(strlen(pstr)-1);
            memcpy(pleaf, pstr+1, strlen(pstr)-2);
            *(pleaf+strlen(pleaf)) = '\0';
            args[0] = strtok( (char *)((*(WATCH_CTX*)arg).cmd), d);
            while(ps = strtok(NULL,d))
            {
                args[j++] = ps;
            }
            args[j++] = strtok(pleaf, d);
            while(ps = strtok(NULL,d))
            {
                args[j++] = ps;
            }

            args[j] = NULL;
        }
        else
        {
            args[0] = strtok( (char *)((*(WATCH_CTX*)arg).cmd), d);
            while(ps = strtok(NULL,d))
            {
                args[j++] = ps;
            }
            
            args[j] = wmalloc(strlen(pstr)+1);
            memcpy(args[j], pstr, strlen(pstr));
            args[++j] = NULL;
        }

        if(strncmp("yes", (*(WATCH_CTX*)arg).verbose, strlen("yes")) == 0)
        {

            printf("exec this cmd and param: ");   
            for(i = 0; i < j; i++)
            {
                printf("%s ", args[i]);
                fflush(stdout);
            }
            printf("\n\n");
            fflush(stdout);
        }
        
        ret = execvp((char *)((*(WATCH_CTX*)arg).cmd), args);  
        wfree(pstr);
        wfree(pleaf);
        for( i = 0; i < j+1; i++)
            wfree(args[i]);
        p_clone->p_wn_key = p_key;
        wilddog_node_delete(p_clone);
        if(ret == -1)
        {
            perror("execv");
            exit(EXIT_FAILURE);
        }
    }
    
    return;
}


int main(int argc, char **argv) 
{
    char url[1024];
    char cmd[256];
    char data[32];
    char verbose[32];
    char quote[32];
    char authvalue[256];
    
    BOOL isFinished = FALSE;
    Wilddog_T wilddog = 0;
    int opt;
    int option_index = 0;
    WATCH_CTX watch_ctx;
    char *host = NULL;
    
    memset(url,0,sizeof(url));  
    memset(cmd,0,sizeof(cmd));
    memset(data,0,sizeof(data));
    memcpy(data, "yes",strlen("yes"));
    memset(verbose,0,sizeof(verbose));
    memcpy(verbose, "no",strlen("no"));
    memset(quote,0,sizeof(quote));
    memcpy(quote, "yes",strlen("yes"));
    memset(authvalue,0,sizeof(authvalue));
	
    static struct option long_options[] = 
    {
        {"ignore-origin-data",     required_argument, 0,  0 },
        {"verbose", required_argument, 0,  0},
        {"ignore-leaf-quote", required_argument, 0,  0},
        {"version", no_argument, 0,  0},
        {"authvalue", required_argument, 0,  0},
        {0,         0,                 0,  0 }
    };

    while ((opt = getopt_long(argc, argv, "ovsh",long_options, &option_index)) != -1) 
    {
        switch (opt) 
        {
        case 0:
            //printf("option!!! %s\n", long_options[option_index].name);
            if (optarg)
            {
                if(strcmp(long_options[option_index].name,"ignore-origin-data")==0)
                    memcpy(data, optarg,strlen(optarg));
                if(strcmp(long_options[option_index].name,"verbose")==0)
                    memcpy(verbose, optarg,strlen(optarg));
                if(strcmp(long_options[option_index].name,"ignore-leaf-quote")==0)
                    memcpy(quote, optarg,strlen(optarg));
                if(strcmp(long_options[option_index].name,"authvalue")==0)
                    memcpy(authvalue, optarg,strlen(optarg));
            }
	    if(strcmp(long_options[option_index].name,"version")==0)
	    {
		    printf("LiveShell Version: %s\n", LIVESHELL_VERSION);	
		    return 0;
	    }
            break;
	case 'o':
	    memcpy(data, "no",strlen("no"));
	    break;
	case 'v':
	    memcpy(verbose, "yes",strlen("yes"));
	    break;
	case 's':
	    memcpy(quote, "no",strlen("no"));
	    break;
		
        case 'h':
            fprintf(stderr, "Usage: %s [option] [--authvalue= your auth data] <your wilddog url>  <your callback command> \n",
                   argv[0]);
            return 0;
        default: /* '?' */
            fprintf(stderr, "Usage: %s [option] [--authvalue= your auth data] <your wilddog url>  <your callback command> \n",
                   argv[0]);
            return 0;
        }
    }
	

    if( argc < 3 )
    {
        printf("Usage: %s [option] [--authvalue= your auth data] <your wilddog url>  <your callback command> \n", argv[0]);
        return 0;
    }
	
    memcpy(cmd, argv[argc-1], strlen(argv[argc-1]));
    memcpy(url, argv[argc-2], strlen(argv[argc-2]));	
	
    wilddog = wilddog_initWithUrl((Wilddog_Str_T *)url);

    if(0 == wilddog)
    {
        wilddog_debug("new wilddog failed!");
        return 0;
    }
    if(strlen(authvalue) != 0)
    {
	    host = getHostFromUrl(url);
	    //printf("host:%s\n", host);
	    //printf("authvalue:%s\n", authvalue);
	    wilddog_auth((u8*)host,(u8*)authvalue,strlen((const char *)authvalue),auth_callback,(void*)&isFinished);
    }
    
    isFinished = FALSE;
    watch_ctx.cmd = cmd;
    watch_ctx.data = data;
    watch_ctx.verbose = verbose;
    watch_ctx.quote = quote;
    watch_ctx.isFinished = &isFinished;
    
    wilddog_addObserver(wilddog, WD_ET_VALUECHANGE, observer_callback, (void*)&watch_ctx);
    while(1)
    {
        if(TRUE == *(watch_ctx.isFinished))
        {
            //wilddog_debug("get new data %d times!", count++);
            *(watch_ctx.isFinished) = FALSE;
            /*do not break*/
            //break;
        }
        wilddog_trySync();
    }
    wfree(host);
    wilddog_destroy(&wilddog);
    
    return 0;
}

